internal/plugin/available_plugins.go

```diff
@@ -11,3 +18,235 @@ type AvailablePlugins interface {
 	HTTPPlugins() map[string]http.Plugin
 	TCPPlugins() map[string]tcp.Plugin
 }
+
+// Plugins represent a holding object for a bundle of plugins of different
+// types.
+type Plugins struct {
+	HTTPPluginMap map[string]http.Plugin
+	TCPPluginMap  map[string]tcp.Plugin
+}
+
+// HTTPPlugins returns only the HTTP plugins in the Plugins struct.
+func (plugins *Plugins) HTTPPlugins() map[string]http.Plugin {
+	return plugins.HTTPPluginMap
+}
+
+// TCPPlugins returns only the TCP plugins in the Plugins struct.
+func (plugins *Plugins) TCPPlugins() map[string]tcp.Plugin {
+	return plugins.TCPPluginMap
+}
+
+type ExternalPluginLookupFunc func(string, string, log.Logger) ([]*go_plugin.Plugin, error)
+type InternalPluginLookupFunc func() (AvailablePlugins, error)
+
+// This method needs to be moved out to the real internal plugin enumeration
+// point.
+func getInternalPluginsFunc() (AvailablePlugins, error) {
+	return &Plugins{}, nil
+}
+
+// This method needs to be moved out to the real external plugin enumeration
+// point.
+func loadPluginsFromDir(pluginDir string,
+	checksumsFile string,
+	logger log.Logger) ([]*go_plugin.Plugin, error) {
+
+	files, err := ioutil.ReadDir(pluginDir)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: Do checksum validation
+	if checksumsFile != "" {
+		// We override file listing if we did a verification to prevent additions
+		// to plugins between verification and loading the plugins.
+		var err error
+		if files, err = VerifyPluginChecksums(pluginDir, checksumsFile); err != nil {
+			logger.Errorln(err)
+			return nil, err
+		}
+	} else if len(files) > 0 {
+		logger.Warnln("Plugin hashes were not provided - tampering will not be detectable!")
+	}
```

Comment:
Here's a piece of cleanup that's not critical (because the above is still pretty clear) but I think is worth doing for extra crispness.  In the above both the nested `if` and the `if... else` break the "line of site" ideal.  In this case, the underlying issue imo is that we're mixing the "load the plugins" code (below this  my comment) and the "deal with checksums" code above it.  That is, we have a SRP violation which is expressing itself in the complexity of those `if`s.  

I'd recommend pulling out the "load the plugins" code into a `loadPluginFiles(files, logger)` method, which does nothing but file loading -- no checksum verification.

Then the code above gets nice and flat:

```go
func loadCheckedPlugins(
	pluginDir string,
	checksumsFile string,
	logger log.Logger,
) ([]*go_plugin.Plugin, error) {

	uncheckedFiles, err := ioutil.ReadDir(pluginDir)
	if err != nil {
		return nil, err
	}

	// no files found, return empty slice
	if len(uncheckedFiles) == 0 {
		return []*go_plugin.Plugin{}, nil
	}

	if checksumsFile == "" {
		logger.Warnln("Plugin hashes were not provided - tampering will not be detectable!")
		return loadPluginFiles(uncheckedFiles, logger), nil
	}

	if checkedFiles, err = VerifyPluginChecksums(pluginDir, checksumsFile); err != nil {
		logger.Errorln(err)
		return nil, err
	}

	return loadPluginFiles(checkedFiles, logger), nil
}

```

Reply:
done